File: scripts/extract_assets.py
#!/usr/bin/env python3
"""
Extrai as imagens dos efeitos da G5 a partir dos arquivos .zrc do ZOOM Edit&Share.

Uso:
    python3 scripts/extract_assets.py "/patchs"

Os .zrc ficam na pasta de dados do Edit&Share (a mesma que tem Module.xml).
As imagens saem em public/fx/<G5_Nome>/*.png e são carregadas pela UI.

Formato do .zrc (engenharia reversa):
  - cabeçalho "ZOOM ZFX  ARCHIVE DATA VER1000"
  - tabela de entradas, um registro a cada 528 bytes, cada um com o nome do arquivo
  - os dados vêm em sequência depois da tabela, PNG atrás de PNG
  A ordem dos nomes .png na tabela é a mesma ordem dos blobs PNG no arquivo.
"""
import os
import re
import sys
import json

PNG_START = re.compile(rb"\x89PNG\r\n\x1a\n")
PNG_END = re.compile(rb"IEND\xaeB\x60\x82")
NAME = re.compile(rb"[A-Za-z0-9_\-]{1,40}\.(?:png|xml)")

OUT = os.path.join(os.path.dirname(__file__), "..", "public", "fx")


def extract(path: str) -> list[str]:
    data = open(path, "rb").read()
    names = [m.group().decode() for m in NAME.finditer(data)]
    pngs = [m.start() for m in PNG_START.finditer(data)]
    ends = [m.start() + 8 for m in PNG_END.finditer(data)]
    png_names = [n for n in names if n.lower().endswith(".png")]

    if not (len(png_names) == len(pngs) == len(ends)):
        print(f"  ! {os.path.basename(path)}: {len(png_names)} nomes / "
              f"{len(pngs)} imagens — pulando")
        return []

    key = os.path.splitext(os.path.basename(path))[0]
    dest = os.path.join(OUT, key)
    os.makedirs(dest, exist_ok=True)
    for name, start, end in zip(png_names, pngs, ends):
        # nomes em minúsculo: o XML referencia "Knob" mas o arquivo é "knob00.png",
        # e URL em servidor web diferencia maiúscula de minúscula
        with open(os.path.join(dest, name.lower()), "wb") as fh:
            fh.write(data[start:end])
    return png_names


def main() -> int:
    if len(sys.argv) < 2:
        print(__doc__)
        return 1
    src = sys.argv[1]
    files = sorted(f for f in os.listdir(src) if f.lower().endswith(".zrc"))
    if not files:
        print(f"Nenhum .zrc em {src}")
        return 1

    manifest: dict[str, list[str]] = {}
    total = 0
    for f in files:
        got = extract(os.path.join(src, f))
        if got:
            manifest[os.path.splitext(f)[0]] = [g.lower() for g in got]
            total += len(got)

    os.makedirs(OUT, exist_ok=True)
    with open(os.path.join(OUT, "manifest.json"), "w") as fh:
        json.dump(manifest, fh, indent=0)
    print(f"{len(manifest)} efeitos, {total} imagens em public/fx/")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/extract_assets.mjs
#!/usr/bin/env node
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const OUT = path.join(__dirname, "..", "public", "fx");

const PNG_SIG  = Buffer.from([0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a]);
const PNG_TAIL = Buffer.from([0x49, 0x45, 0x4e, 0x44, 0xae, 0x42, 0x60, 0x82]);
const NAME_RE  = /[A-Za-z0-9_-]{1,40}\.(?:png|xml)/gi;

function findAll(buf, pattern) {
  const positions = [];
  let pos = 0;
  while (true) {
    const idx = buf.indexOf(pattern, pos);
    if (idx === -1) break;
    positions.push(idx);
    pos = idx + 1;
  }
  return positions;
}

function extract(filePath) {
  const data = fs.readFileSync(filePath);
  NAME_RE.lastIndex = 0;
  const allNames = [...data.toString("latin1").matchAll(NAME_RE)].map(m => m[0]);
  const pngNames = allNames.filter(n => n.toLowerCase().endsWith(".png"));
  const pngStarts = findAll(data, PNG_SIG);
  const pngEnds   = findAll(data, PNG_TAIL).map(p => p + 8);

  if (pngNames.length !== pngStarts.length || pngNames.length !== pngEnds.length) {
    console.log(`  ! ${path.basename(filePath)}: ${pngNames.length} nomes / ${pngStarts.length} imagens — pulando`);
    return [];
  }

  const key  = path.basename(filePath, ".zrc");
  const dest = path.join(OUT, key);
  fs.mkdirSync(dest, { recursive: true });

  for (let i = 0; i < pngNames.length; i++) {
    const name = pngNames[i].toLowerCase();
    fs.writeFileSync(path.join(dest, name), data.subarray(pngStarts[i], pngEnds[i]));
  }
  return pngNames;
}

function walkZrc(dir) {
  const results = [];
  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
    const full = path.join(dir, entry.name);
    if (entry.isDirectory()) results.push(...walkZrc(full));
    else if (entry.name.toLowerCase().endsWith(".zrc")) results.push(full);
  }
  return results.sort();
}

const flags = new Set(process.argv.slice(2).filter(a => a.startsWith("--")));
const src   = process.argv.slice(2).find(a => !a.startsWith("--"));

if (!src) {
  console.error("Uso: node scripts/extract_assets.mjs <pasta> [--if-missing]");
  process.exit(1);
}

if (flags.has("--if-missing") && fs.existsSync(path.join(OUT, "manifest.json"))) {
  process.exit(0);
}

const files = walkZrc(src);
if (!files.length) {
  console.error(`Nenhum .zrc em ${src}`);
  process.exit(1);
}

const manifest = {};
let total = 0;
for (const file of files) {
  const got = extract(file);
  if (got.length) {
    manifest[path.basename(file, ".zrc")] = got.map(n => n.toLowerCase());
    total += got.length;
  }
}

fs.mkdirSync(OUT, { recursive: true });
fs.writeFileSync(path.join(OUT, "manifest.json"), JSON.stringify(manifest, null, 0));
console.log(`${Object.keys(manifest).length} efeitos, ${total} imagens → public/fx/`);
